packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,100 @@
+import {auth, nwc, webln} from '@getalby/sdk';
+import { useEffect, useState } from "react";
+import { getUmaDomain } from "src/utils/getUmaDomain";
+import { useUser } from "./useUser";
+import { create } from 'zustand';
+import { persist } from "zustand/middleware";
+import { getNwcUmaDomain } from 'src/utils/getNwcUmaDomain';
+
+type UmaAuthToken = {
+  access_token: string,
+  refresh_token: string,
+  expires_at: number,
+  nwc_connection_uri: string;
+  commands: string[];
+  budget: string;
+  nwc_expires_at: string;
+};
+
+interface TokenState {
+  accessToken: string;
+  refreshToken: string;
+  expiresAt: number;
+}
+
+interface OAuthState {
+  codeVerifier?: string;
+  token?: TokenState;
+  identityNpub?: string;
+  identityRelayUrl?: string;
+  redirectUri?: string;
+  nwcConnectionUri?: string;
+  setToken: (token: TokenState) => void;
+  performOAuth: (uma: string) => Promise<void>;
+}
+
+export const useOAuth = create<OAuthState>()(
+  persist(
+    (set, get) => ({
+      setToken: (token) => set({
+        token,
+      }),
+      performOAuth: async (uma) => {
+        const state = get();
+        const result = await performOAuth(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+      })
+    }
+  )
+);
+
+const performOAuth = async (state: OAuthState, uma: string) => {
+  const { identityNpub, identityRelayUrl, redirectUri, codeVerifier, token, nwcConnectionUri } = state;
+
+  // If we have a connection URI and a token that hasn't expired, we don't need to do anything
+  if (nwcConnectionUri && token && (!token.expiresAt || token.expiresAt < Date.now())) {
+    return state;
+  }
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!, // TODO: get this from client app
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: getNwcUmaDomain(uma),
+    },
+  });
+
+  let resultToken: UmaAuthToken;
+  if (token?.refreshToken) {
+    authClient.token = {
+      access_token: token.accessToken,
+      refresh_token: token.refreshToken,
+      expires_at: token.expiresAt,
+    };
+
+    resultToken = (await authClient.refreshAccessToken()).token as UmaAuthToken;
+  } else {
+    authClient.code_verifier = codeVerifier!; // TODO: where do we get this?
```

Comment:
This is generated by OAuthUser during the auth request and should then be saved locally so that it can be pulled back out on the redirect. This is where I [save it in bitcoin-connect](https://github.com/jklein24/bitcoin-connect/blob/c3de26d3ae2799d5b5772811b4b39f3889009319/src/components/pages/bc-uma.ts#L83).

Reply:
Oh i was referring to https://github.com/jklein24/bitcoin-connect/blob/feat/crosscurrency/src/connectors/UMANWCConnector.ts#L55 and was wondering where it was coming from... by saving locally you mean in localStorage? We'd need to persist it between redirects right?

Reply:
In that code the code_verifier always gets reset to an empty string

Reply:
Yep - you need to save it before starting the oauth flow so that it's there when you come back and need to do the token exchange.

1. Generate code_challenge + verifier. Save them somewhere.
2. Do the authorization request/redirect
3. You get redirected back with code and state.
4. You pull the code and verifier to use in the token exchange request.